### PR TITLE
BestFirstSearch: check optimal token not to overflow

### PR DESCRIPTION
**Pull request by kitbellew**
Follow the post-optimal states on the same line while we get exactly one active split to use; if the line overflows, don't mark optimal as found. This way, we are more likely to fit within the column limit and select a better solution.

Also, now that BestFirstSearch logic dealing with optimal tokens has been improved, the suboptimal special case handling of infix expressions around blocks is no longer needed.

`scala-repos` diffs:
* BestFirstSearch: check optimal tok not to overflow: https://github.com/kitbellew/scala-repos/commit/fd8b18e8d4638c73724a343b343a9e9c83151c00?w=1
* Router: always use optimal token within infix op: https://github.com/kitbellew/scala-repos/commit/5a8132dc0f796690190079d2b8be6fd5d01398c1?w=1

### CODE REVIEW

**Comment by kitbellew** [comment]
@olafurpg @tanishiking @poslegm PTAL. another PR to combat line overflow, mostly around infix expressions.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala [hunk 1]
```diff
@@ -215,14 +202,20 @@ private class BestFirstSearch private (
                     nextState.split.cost == 0 =>
                 val nextNextState =
                   shortestPath(nextState, token, depth + 1, maxCost = 0)
-                if (null != nextNextState) {
-                  optimalNotFound = false
-//                  logger.elem(split, splitToken, formatWriter.mkString(nextNextState.splits), tokens(nextNextState.splits.length))
-                  Q.enqueue(nextNextState)
+                val furtherState =
+                  if (null == nextNextState) nextNextState
+                  else if (!style.activeForEdition_2020_03) nextNextState
```
Comment (poslegm) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala]:
```suggestion
                  if (null == nextNextState || !style.activeForEdition_2020_03) nextNextState
```

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala [hunk 2]
```diff
@@ -251,6 +244,50 @@ private class BestFirstSearch private (
     null
   }
 
+  private def getActiveSplits(state: State, ft: FormatToken): Seq[Split] = {
+    val splits =
+      if (state.formatOff) List(provided(ft))
+      else if (ft.inside(range)) routes(state.depth)
+      else List(provided(ft))
```
Comment (poslegm) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala]:
```suggestion
      if (ft.inside(range)) routes(state.depth)
      else List(provided(ft))
```

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala [hunk 3]
```diff
@@ -251,6 +244,50 @@ private class BestFirstSearch private (
...
+  @tailrec
+  private def traverseSingleSplitPath(state: State, depth: Int): State =
+    if (state.depth >= tokens.length) state
+    else {
+      trackState(state, depth, 0)
+
+      val splitToken = tokens(state.depth)
+      val activeSplits = getActiveSplits(state, splitToken)
+
+      if (activeSplits.length != 1)
```
Comment (poslegm) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala]:
```suggestion
      if (activeSplits.lengthCompare(1) != 0)
```

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala [hunk 4]
```diff
@@ -251,6 +244,50 @@ private class BestFirstSearch private (
...
+        if (activeSplits.length != 0) state else null // dead end
```
Comment (poslegm) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/BestFirstSearch.scala]:
```suggestion
        if (activeSplits.nonEmpty) state else null // dead end
```

**Pull request merged**